src/Hl7.Fhir.Support.Poco/Model/FhirDateTime.cs

```diff
@@ -107,6 +107,77 @@ public DateTimeOffset ToDateTimeOffset(TimeSpan zone)
         }
 
 
+        /// <summary>
+        /// Determines whether a Fhir DateTime can be converted to a a .NET DateTimeOffset. Fails in case of a partial DateTime (no timezone), 
+        /// otherwise 'success' will be returned including the DateTimeOffset as out parameter. 
+        /// </summary>
+        /// <param name="dto">A DateTimeOffset, if a timezone is correctly specified</param>
+        /// <returns>A boolean representing if a DateTimeOffset keeping the original timezone can be converted</returns>
+        public bool TryToDateTimeOffset(out DateTimeOffset dto)
+        {
+            dto = default;
+
+            if (this.Value == null)
+                return false;          
+
+            if (tryGetTimeSpan(out var timespan))
+            {
+                dto = ToDateTimeOffset(timespan);
+                return true;
+            }
+            else
+            {
+                return false;
+            }           
+        }
+
+        private bool tryGetTimeSpan(out TimeSpan ts)
+        {
+            ts = default;
+
+            if (this.containsTimeZone())
+            {
+                if (this.Value.Substring(this.Value.Length - 1) == "Z")
+                {
+                    ts = TimeSpan.Zero;
+                    return true;
+                }
+                
+                //converts +07:00 to 07:00, and keeps -07:00 as is.
+                var tz = this.Value.Substring(this.Value.Length - 6).Replace("+", "");
+
+                if (TimeSpan.TryParse(tz, out var timespan))
+                {
+                    ts = timespan;
+                    return true;
+                }                    
+                else
+                {
+                    //can't determine timezone
+                    return false;
+                }                    
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        private bool containsTimeZone()
+        {
+            //YYYY-MM-DDThh:mm:ss.sss+zz:zz // 
```

Comment:
The regex of a `FhirDateTime` is:
```
([0-9]([0-9]([0-9][1-9]|[1-9]0)|[1-9]00)|[1-9]000)(-(0[1-9]|1[0-2])(-(0[1-9]|[1-2][0-9]|3[0-1])(T([01][0-9]|2[0-3]):[0-5][0-9]:([0-5][0-9]|60)(\.[0-9]+)?(Z|(\+|-)((0[0-9]|1[0-3]):[0-5][0-9]|14:00)))?)?)?
```

So the fraction after the second, can be very precise. For example this is also a correct FhirDateTime:

```
2021-03-18T12:22:35.12345+04:00
```

So the `+` or `-` of the Timezone is not always on position 23 in the string.

Would it not be an idea to parse the string with the regex? Then you can extract the timezone from that.

Reply:
You can get some inspiration from this code: https://github.com/FirelyTeam/firely-net-common/blob/115fce4c1fab236fd80685aa1030ba82be925edd/src/Hl7.Fhir.ElementModel/Types/DateTime.cs#L93